juju/juju
# Hook up the deployer nested context.

This work introduces the implementation of the unified machine
and unit agent for IAAS models. The deployer now runs the workers
for the units in the same process using a runner to run a primary
worker for each unit. This primary worker is a dependency engine
that runs a subset of the workers that the old unit agent ran.
There is no need for the units to now run upgrade workers as
that is all handled by the machine agent.

An upgrade step is added to disable all the service files running
for the old unit agents, and update the agent.conf file for the
machine agent to record the deployed units.

## QA steps

```sh
juju bootstrap lxd test
juju deploy cs:~jameinel/ubuntu-lite ubuntu
juju deploy filebeat
juju relate ubuntu filebeat
watch -c juju status --color
```
Wait until the units are deployed.
```sh
juju ssh 0
```
See that the file structure in `/var/lib/juju` is unchanged. Look at the `juju_engine_report` and observe that the `deployer` now shows the units as deployed, and the nested engine report is shown for each of the units.
Also note that `/etc/systemd/system` only has a machine agent, and no unit agents.

Upgrade steps testing
```sh
sudo snap refresh juju --channel 2.8
/snap/bin/juju bootstrap lxd upgrade --no-gui
juju deploy cs:~jameinel/ubuntu-lite ubuntu
juju deploy filebeat
juju relate ubuntu filebeat
watch -c juju status --color
```
Wait until deployed, then
```sh
juju upgrade-controller --build-agent
# wait a bit for the controller to upgrade
juju upgrade-model --agent-version 2.9-beta1.1
```
then
```sh
juju ssh 0
```
Check `/etc/systemd/system` to observe that no unit agents exist, then run `juju_engine_report` to
observe that the unit workers are running in the machine agent.


## Documentation changes

We probably want to update the docs to remove references to the unit agents as separate entities for IAAS models.
